docs/enhancements/token-exchange-2023-02-03-#2812.md

```diff
@@ -0,0 +1,171 @@
+[#1225 GitHub Non-Web application flow support]: https://github.com/dexidp/dex/issues/1225
+
+An initial attempt is at [#2806](https://github.com/dexidp/dex/pull/2806)
+
+## Motivation
+
+### Goals/Pain
+
+The goal is to allow programmatic access to Dex-protected resources 
+without the use of static/long-lived secret tokens (API keys, username/password)
+or web-based redirect flows.
+Such scenarios are common in CI/CD workflows,
+and in general automation of common tasks.
+
+### Non-goals
+
+- Work will be scoped to just the OIDC connector
+- [RFC 8693 Section 2.1.1. Relationship between Resource, Audience, and Scope]
+  details more complex authorization checks based on targeted resources.
+  This is considered out of scope.
+
+[RFC 8693 Section 2.1.1. Relationship between Resource, Audience, and Scope]: https://www.rfc-editor.org/rfc/rfc8693.html#name-relationship-between-resour
+
+## Proposal
+
+### User Experience
+
+Clients can make `POST` requests with `application/x-www-form-urlencoded` 
+parameters as specified by [RFC 8693] to Dex's `/token` endpoint.
+If successful, an access token will be returned,
+allowing direct authentication with Dex.
+No refresh tokens will be issued,
+perform a new exchange (possibly with refreshed upstream tokens) to obtain a new access token.
+
+The request parameters from [RFC 8693 Section 2.1](https://www.rfc-editor.org/rfc/rfc8693.html#name-request):
+
+- `grant_type`: REQUIRED - `urn:ietf:params:oauth:grant-type:token-exchange`
+- `resource`: OPTIONAL - the `audience` in the issued Dex token
+- `audience`: REQUIRED (RFC OPTIONAL) - the connector to verify the provided token against
+- `scope`: OPTIONAL - the `scope` in the issued Dex token
+- `requested_token_type`: OPTIONAL - one of `urn:ietf:params:oauth:token-type:access_token` or `urn:ietf:params:oauth:token-type:id_token`, defaulting to access token
+- `subject_token`: REQUIRED - the token issued by the upstream IDP
+- `subject_token_type`: REQUIRED - only  `urn:ietf:params:oauth:token-type:id_token` will be accepted
+- `actor_token`: OPTIONAL - unused
+- `actor_token_type`: OPTIONAL - unused
+
+The response parameters from [RFC 8693 Section 2.2](https://www.rfc-editor.org/rfc/rfc8693.html#name-response):
+
+- `access_token`: the issued token, the field is called `access_token` for legacy reasons
+- `issued_token_type`: the actual type of the issued token
+- `token_type`: the value `Bearer`
+- `expires_in`: validity lifetime in seconds
+- `scope`: the requested scope
+- `refresh_token`: unused
+
+The connector only needs to be configured with an issuer,
+no client ID / client secrets are necessary
+
+```yaml
+connectors:
+- type: oidc
+  id: my-platform
+  name: My Platform
+  config:
+    issuer: https://oidc.my-platform.example/
+```
+
+### Implementation Details/Notes/Constraints
+
+- Connectors expose a new interface `TokenIdentity` that will verify the given token and return the associated identity.
```

Comment:
This is controversial because it adds more entropy to the connector's interface.

Dex has three connector types (four, to be precise, but let's forget about SAML for now) - Password, Callback, and Refresh. It makes seven possible combinations of connector functions. With an additional interface, the number will be fourteen. 

However, thinking about alternatives, I like the solution of exposing an additional interface. In the future, a JWT connector can be mentioned in this DEP that will implement only the TokenIdentity interface.

Reply:
Yes, I was initially hesitant to expose an additional interface, but felt that this was the cleanest way to expose the logic (I started with inlining the verify call in the handler)